rfcs/0000-dsl-driven-rda-code-generation.md

```diff
@@ -0,0 +1,681 @@
+- The `random-data` goal generates random data generation classes to create randomized data of appropriate size and type for each object/field in the RDA API messages.
+- The `synthea-bridge` goal generates data transformation classes to copy data from Synthea RIF data files into protobuf messages.
+
+For an idea of the code savings consider the difference in complexity between that YAML example and these hand written classes:
+
+- [PreAdjFissClaims hand written entity](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-model/bfd-model-rda/src/main/java/gov/cms/bfd/model/rda/PreAdjFissClaim.java)
+- [FissClaimTransformer hand written transformer class](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/FissClaimTransformer.java)
+- [RandomFissClaimGenerator hand written synthetic data class](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/RandomFissClaimGenerator.java)
+
+Getting the relationships right between tables in JPA can be somewhat tricky.
+However, they can be trivially defined as `array`s in the YAML and the code generator takes care of getting the details correct:
+
+````yaml
+  - id: FissClaim
+    messageClassName: gov.cms.mpsm.rda.v1.fiss.FissClaim
+    entityClassName: gov.cms.bfd.model.rda.PreAdjFissClaim
+    transformerClassName: gov.cms.bfd.pipeline.rda.grpc.source.FissClaimTransformer2
+    table:
+      name: FissClaims
+      schema: pre_adj
+      primaryKeyColumns:
+        - dcn
+      columns:
+        - name: dcn
+          sqlType: varchar(23)
+          nullable: false
+        - name: sequenceNumber
+          sqlType: bigint
+          nullable: false
+        - name: hicNo
+          sqlType: varchar(12)
+          nullable: false
+        # ... skipping lots of columns for the sake of this example ...
+    transformations:
+      - from: dcn
+        optional: false
+      - from: seq
+        to: sequenceNumber
+        optional: false
+      - from: hicNo
+        optional: false
+      # ... skipping lots of transforms for the sake of this example ...
+    arrays:
+      - from: fissProcCodes
+        to: procCodes
+        mapping: FissProcCode
+      - from: fissDiagCodes
+        to: diagCodes
+        mapping: FissDiagnosisCode
+      - from: fissPayers
+        to: payers
+        mapping: FissPayer
+        namePrefix: "payer"
+````
+
+The example illustrates the three detail tables associated with each `FissClaim`.
+The RDA API sends these as `repeated` fields in the protobuf definition and the plugin maps them to detail entities in the JPA classes.
+Each one references the field in the protobuf message and the entity class and the mapping used to define the detail table.
+
+The classes generated by the plugin rely on a few utility classes that are defined in a separate library module.
```

Comment:
I might consider moving this section into the details section below.  It's getting into potential implementation details of how to interpret the DSL, and I think detracts a bit from the DSL itself.  Good info, just I don't think it's primary to the proposal.

Reply:
Hmm, I like having it here in the motivation section though since it illustrates the kind of thing that's easy to get wrong when doing things by hand.